doc/reference/kernel/other/fatal.rst

```diff
@@ -0,0 +1,245 @@
+.. _fatal:
+
+Fatal Errors
+############
+
+Software Errors Triggered in Source Code
+****************************************
+
+Zephyr provides several methods for inducing fatal error conditions through
+either build-time checks, conditionally compiled assertions, or deliberately
+invoked panic or oops conditions.
+
+Runtime Assertions
+==================
+
+Zephyr provides some macros to perform runtime assertions which may be
+conditionally compiled. Their defintions may be found in
+``include/sys/__assert.h``.
+
+Assertions are enabled by setting the ``__ASSERT_ON`` preprocessor symbol to a
+non-zero value. There are two ways to do this:
+
+- Use the ``CONFIG_ASSERT`` and ``CONFIG_ASSERT_LEVEL`` kconfig options.
+- Add ``-D__ASSERT_ON=<level>`` to the project's CFLAGS, either on the
+  build command line or in a CMakeLists.txt.
+
+The ``__ASSERT_ON`` method takes precedence over the kconfig option if both are
+used.
+
+Specifying an assertion level of 1 causes the compiler to issue warnings that
+the kernel contains debug-type ``__ASSERT()`` statements; this reminder is
+issued since assertion code is not normally present in a final product.
+Specifying assertion level 2 suppresses these warnings.
+
+Assertions are enabled by default when running Zephyr test cases, which
+all enable ``CONFIG_TEST``.
+
+The policy for what to do when encountering a failed assertion is controlled
+by the implementation of :c:func:`assert_post_action`. Zephyr provides
+a default implementation with weak linkage which invokes a kernel oops if
+the thread that failed the assertion was running in user mode, and a kernel
+panic otherwise.
```

Comment:
`subsys/debug/Kconfig` has what appears to be an outdated description of the behavior of `__ASSERT` ("...calling thread is put on an infinite tight loop") that should be removed.

Reply:
fixed